plugins/askrene/askrene.c

```diff
@@ -0,0 +1,915 @@
+
+	layer = find_layer(askrene, layername);
+	if (!layer)
+		layer = new_layer(askrene, layername);
+
+	/* We save this in the layer, because they want us to disable all the channels
+	 * to the node at *use* time (a new channel might be gossiped!). */
+	layer_add_disabled_node(layer, node);
+
+	response = jsonrpc_stream_success(cmd);
+	return command_finished(cmd, response);
+}
+
+static struct command_result *json_askrene_listlayers(struct command *cmd,
+						      const char *buffer,
+						      const jsmntok_t *params)
+{
+	struct askrene *askrene = get_askrene(cmd->plugin);
+	const char *layername;
+	struct json_stream *response;
+
+	if (!param(cmd, buffer, params,
+		   p_opt("layer", param_string, &layername),
+		   NULL))
+		return command_param_failed();
+
+	response = jsonrpc_stream_success(cmd);
+	json_add_layers(response, askrene, "layers", layername);
+	return command_finished(cmd, response);
+}
+
+static struct command_result *json_askrene_age(struct command *cmd,
+						 const char *buffer,
+						 const jsmntok_t *params)
+{
+	struct layer *layer;
+	struct json_stream *response;
+	u64 *cutoff;
+	size_t num_removed;
+
+	if (!param(cmd, buffer, params,
+		   p_req("layer", param_known_layer, &layer),
+		   p_req("cutoff", param_u64, &cutoff),
+		   NULL))
+		return command_param_failed();
+
+	num_removed = layer_trim_constraints(layer, *cutoff);
+
+	response = jsonrpc_stream_success(cmd);
+	json_add_string(response, "layer", layer_name(layer));
+	json_add_u64(response, "num_removed", num_removed);
+	return command_finished(cmd, response);
+}
+
+static const struct plugin_command commands[] = {
+	{
+		"getroutes",
+		json_getroutes,
+	},
+	{
+		"askrene-reserve",
+		json_askrene_reserve,
+	},
+	{
+		"askrene-unreserve",
+		json_askrene_unreserve,
+	},
+	{
+		"askrene-disable-node",
+		json_askrene_disable_node,
+	},
+	{
+		"askrene-create-channel",
+		json_askrene_create_channel,
+	},
+	{
+		"askrene-inform-channel",
+		json_askrene_inform_channel,
+	},
+	{
+		"askrene-listlayers",
+		json_askrene_listlayers,
+	},
+	{
+		"askrene-age",
+		json_askrene_age,
+	},
+};
+
+static void askrene_markmem(struct plugin *plugin, struct htable *memtable)
+{
+	layer_memleak_mark(get_askrene(plugin), memtable);
+}
+
+static const char *init(struct plugin *plugin,
+			const char *buf UNUSED, const jsmntok_t *config UNUSED)
+{
+	struct askrene *askrene = tal(plugin, struct askrene);
+	askrene->plugin = plugin;
+	list_head_init(&askrene->layers);
+	askrene->reserved = new_reserve_hash(askrene);
+	askrene->gossmap = gossmap_load(askrene, GOSSIP_STORE_FILENAME, NULL);
+
+	if (!askrene->gossmap)
+		plugin_err(plugin, "Could not load gossmap %s: %s",
+			   GOSSIP_STORE_FILENAME, strerror(errno));
+	askrene->capacities = get_capacities(askrene, askrene->plugin, askrene->gossmap);
+	rpc_scan(plugin, "getinfo", take(json_out_obj(NULL, NULL, NULL)),
+		 "{id:%}", JSON_SCAN(json_to_node_id, &askrene->my_id));
+
+	plugin_set_data(plugin, askrene);
```

Comment:
Instead of calling `plugin_set_data` here, why not pass the data as argument to `plugin_main`?

It is not clear to me what happens if I put that set data statement in `init` but also pass a data pointer in `plugin_main`. One of the two would come last and override the first. How can we avoid the user shooting himself on the foot?

Reply:
It would override if you did both.

But the problem is that you should really do setup before allocating, so if you want to allocate it's awkward to do it before calling plugin_main.  For example, leak detection setup.

Perhaps the *main* argument was the error, and we should have used this in init for all plugins?  Any significant plugin has a main function anyway?